### PR TITLE
[3621, 3651] Afford add/remove individual component to/from pools

### PR DESCRIPTION
**Pull request by wwelling**
**[VIVO GitHub issue](https://github.com/vivo-project/VIVO/issues)**: 
https://github.com/vivo-project/VIVO/issues/3621
https://github.com/vivo-project/VIVO/issues/3621

# What does this pull request do?
Add and remove methods added to abstract pool. Before adding ensures model has component URI and before removing ensures model no longer has component URI.

# What's new?
Add and remove methods on pool interface and implemented in abstract pool.
```
public void add(String uri, C component);

public void remove(String uri, String name);
```

# Interested parties
Tag (@ mention) interested parties or, if unsure, @VIVO-project/vivo-committers

### CODE REVIEW

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java [hunk 1]
```diff
@@ -22,6 +25,10 @@
 
     private final Log log = LogFactory.getLog(this.getClass());
 
+    private final static String RDF_TYPE_PROPERTY = "http://www.w3.org/1999/02/22-rdf-syntax-ns#type";
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java:28]: https://github.com/vivo-project/Vitro/blob/39e1eed20df592e8c0b62834d34042bc4d370ae3/api/src/main/java/edu/cornell/mannlib/vitro/webapp/dao/VitroVocabulary.java#L29

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java [hunk 2]
```diff
@@ -126,4 +162,11 @@ private boolean isComponentInUse(C component) {
         return true;
     }
 
+    private boolean isInModel(String uri) {
+        Resource s = dynamicAPIModel.getResource(uri);
+        Property p = dynamicAPIModel.getProperty(RDF_TYPE_PROPERTY); 
+
+        return dynamicAPIModel.contains(s, p);
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java:169]: Is there an option to verify that object type is right (not only subject and predicate)?

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java (hunk 1 above)
Comment (wwelling) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java:28]: Thanks. Will use an existing constant.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ResourcePool.java [hunk 3]
```diff
@@ -26,4 +28,9 @@ public Resource getDefault() {
 		return Resource.class;
 	}
 
+	@Override
+	public String getRDFType() {
+		return DYNAPI_RESOURCE;
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ResourcePool.java:33]: The check needs to be against java uri, not ontology class uri (example java:edu.cornell.mannlib.vitro.webapp.dynapi.components.Action)
Can be done with this method
https://github.com/vivo-project/Vitro/blob/39e1eed20df592e8c0b62834d34042bc4d370ae3/api/src/main/java/edu/cornell/mannlib/vitro/webapp/utils/configuration/ConfigurationBeanLoader.java#L34

**Review by chenejac — APPROVED** [review]
Reloading one action/resource will be realized by invoking remove and add, right?

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ResourcePool.java (hunk 3 above)
Comment (wwelling) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ResourcePool.java:33]: Not working.

```
private boolean isInModel(String uri) {
        Resource s = dynamicAPIModel.getResource(uri);
        Property p = dynamicAPIModel.getProperty(RDF_TYPE); 
        ObjectProperty o = dynamicAPIModel.getObjectProperty(toJavaUri());

        StmtIterator it = dynamicAPIModel.listStatements();
        while(it.hasNext()) {
            Statement st = it.next();
            if (st.toString().contains(uri) &&  st.toString().contains(toJavaUri())) {
                System.out.println("Found: ");
                System.out.println(st);
            }
        }

        System.out.println("\n" + toJavaUri() + " => " + dynamicAPIModel.getObjectProperty(toJavaUri()) + "\n");
        System.out.println("\n" + ConfigurationBeanLoader.toJavaUri(getType()) + " => " + dynamicAPIModel.getObjectProperty(ConfigurationBeanLoader.toJavaUri(getType())) + "\n");

        System.out.println("\n" + s + ", " + p + ", " + toJavaUri() + " => " + dynamicAPIModel.contains(s, p, toJavaUri()) + "\n");
        System.out.println("\n" + s + ", " + p + ", " + ConfigurationBeanLoader.toJavaUri(getType()) + " => " + dynamicAPIModel.contains(s, p, ConfigurationBeanLoader.toJavaUri(getType())) + "\n");

        return dynamicAPIModel.contains(s, p, toJavaUri());
    }

    private String toJavaUri() {
        Class<C> clazz = getType();
        String javaUri = ConfigurationBeanLoader.toJavaUri(clazz);
        String simpleName = clazz.getSimpleName();

        return javaUri.replace(format(".%s", simpleName), format("#%s", simpleName));
    }
 ```
 
 ```
 Found:
[https://vivoweb.org/ontology/vitro-dynamic-api/action/testReloadAction1, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, java:edu.cornell.mannlib.vitro.webapp.dynapi.components#Action]

java:edu.cornell.mannlib.vitro.webapp.dynapi.components#Action => null


java:edu.cornell.mannlib.vitro.webapp.dynapi.components.Action => null


https://vivoweb.org/ontology/vitro-dynamic-api/action/testReloadAction1, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, java:edu.cornell.mannlib.vitro.webapp.dynapi.components#Action => false


https://vivoweb.org/ontology/vitro-dynamic-api/action/testReloadAction1, http://www.w3.org/1999/02/22-rdf-syntax-ns#type, java:edu.cornell.mannlib.vitro.webapp.dynapi.components.Action => false
```

Comment (wwelling) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ResourcePool.java:33]: @litvinovg thoughts?

Comment (wwelling) [reply api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/ResourcePool.java]: @litvinovg I have implemented by listing object properties for the resource and property and checking if any match the expected Java URI.

**Comment by wwelling** [comment]
@chenejac I would suggest we add a `reload(String uri)` to reload single action/resource. Load and reload are pulled directly from the model and do not require inputs. Remove and add would require getting the component from the model outside of the pool. I will add reload single to this PR.

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java [hunk 4]
```diff
@@ -52,17 +62,54 @@ public C getByName(String name) {
...
+            throw new RuntimeException(format("%s %s with URI %s not found in model. Not adding to pool.", getType().getName(), name, uri));
+        }
+    }
+
+    public void remove(String uri, String name) {
+        log.info(format("Removing component %s with URI %s", name, uri));
+        if (!isInModel(uri)) {
+            synchronized (mutex) {
+                C oldComponent = components.remove(name);
+                if (oldComponent != null && oldComponent.hasClients()) {
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java:89]: Maybe it is better to put object to obsoleteComponents and run unloadObsoleteComponents?

File: api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java [hunk 5]
```diff
@@ -52,17 +62,54 @@ public C getByName(String name) {
...
+        }
+    }
+
+    public void add(String uri, C component) {
+        String name = component.getName();
+        log.info(format("Adding component %s with URI %s", name, uri));
+        if (isInModel(uri)) {
+            synchronized (mutex) {
+                C oldComponent = components.put(name, component);
+                if (oldComponent != null && oldComponent.hasClients()) {
```
Comment (litvinovg) [inline api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/AbstractPool.java:75]: The same, I would add it to obsolete components and let it check if it has clients and run dereferencing

**Review by litvinovg — APPROVED** [review]
Looks good to me

**Pull request merged**